### PR TITLE
fix: Fix the CUDAGraphs C++ runtime implementation

### PR DESCRIPTION
**Pull request by narendasan**
# Description
Fixes the implementation intra-engine CUDAGraph support in the C++ runtime. Cleans up the implementation to distinctly handle creating and maintaining persistent buffers as well as adds explicit concurrency management between the calling stream and the engine execution stream. 

Fixes #

## Type of change

Please delete options that are not relevant and/or add your own.

- Bug fix (non-breaking change which fixes an issue)

# Checklist:

- [x] My code follows the style guidelines of this project (You can use the linters)
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas and hacks
- [x] I have made corresponding changes to the documentation
- [x] I have added tests to verify my fix or my feature
- [x] New and existing unit tests pass locally with my changes
- [x] I have added the relevant labels to my PR in so that relevant reviewers are notified

### CODE REVIEW

**Review by narendasan — COMMENTED** [review]
image

File: .github/workflows/build-test-windows.yml [hunk 1]
```diff
@@ -219,7 +219,7 @@ jobs:
         export USE_HOST_DEPS=1
         pushd .
         cd tests/py/dynamo
-        python -m pytest -n 4 --junitxml=${RUNNER_TEST_RESULTS_DIR}/tests_py_dynamo_core_runtime_test_results.xml runtime/
+        python -m pytest --junitxml=${RUNNER_TEST_RESULTS_DIR}/tests_py_dynamo_core_runtime_test_results.xml runtime/
```
Comment (narendasan) [inline .github/workflows/build-test-windows.yml:222]: Update windows tests as well

**Review by zewenli98 — APPROVED** [review]
LGTM

**Review by peri044 — APPROVED** [review]

File: .github/workflows/build-test-linux.yml [hunk 2]
```diff
@@ -229,9 +229,39 @@ jobs:
...
+      ref: ""
+      test-infra-repository: pytorch/test-infra
+      test-infra-ref: main
+      build-matrix: ${{ needs.generate-matrix.outputs.matrix }}
+      pre-script: ${{ matrix.pre-script }}
+      script: |
+        export USE_HOST_DEPS=1
+        pushd .
+        cd tests/py/dynamo
+        nvidia-smi
```
[the remark this answers is not in the document]
Comment (narendasan) [reply .github/workflows/build-test-linux.yml:262]: Yeah

**Pull request merged**